pinot-core/src/main/java/com/linkedin/pinot/core/operator/transform/DateTimeConversionEvaluator.java

```diff
@@ -0,0 +1,213 @@
+
+import java.util.concurrent.TimeUnit;
+
+import org.apache.commons.lang3.EnumUtils;
+import org.joda.time.DateTime;
+import org.joda.time.DateTimeZone;
+import org.joda.time.Days;
+import org.joda.time.Hours;
+import org.joda.time.Interval;
+import org.joda.time.Minutes;
+import org.joda.time.Months;
+import org.joda.time.MutableDateTime;
+import org.joda.time.Seconds;
+import org.joda.time.Weeks;
+import org.joda.time.Years;
+import org.joda.time.format.DateTimeFormat;
+import org.joda.time.format.DateTimeFormatter;
+
+import com.google.common.base.Preconditions;
+import com.linkedin.pinot.common.data.DateTimeFieldSpec.TimeFormat;
+import com.linkedin.pinot.common.utils.time.DateTimeFieldSpecUtils;
+
+/**
+ * Converts the time column value based on the inputFormat, outputFormat and outputGranularity
+ */
+public class DateTimeConversionEvaluator {
+
+  private static final DateTime EPOCH_START_DATE = new DateTime(0L, DateTimeZone.UTC);
+
+  public enum DateTimeTransformUnit {
+    YEARS,
+    MONTHS,
+    WEEKS,
+    DAYS,
+    HOURS,
+    MINUTES,
+    SECONDS,
+    MILLISECONDS;
+  }
+
+  private String inputFormat;
+  private String outputFormat;
+  private String outputGranularity;
+  private DateTimeFormatter inputDateTimeFormatter = null;
+  private DateTimeFormatter outputDateTimeFormatter = null;
+
+  public DateTimeConversionEvaluator(String inputFormat, String outputFormat, String outputGranularity) {
+
+    Preconditions.checkArgument(DateTimeFieldSpecUtils.isValidFormat(inputFormat));
+    Preconditions.checkArgument(isValidFormat(outputFormat));
+    Preconditions.checkArgument(DateTimeFieldSpecUtils.isValidGranularity(outputGranularity));
+    this.inputFormat = inputFormat;
+    this.outputFormat = outputFormat;
+    this.outputGranularity = outputGranularity;
+    String inputSDFFormat = DateTimeFieldSpecUtils.getSDFPatternFromFormat(inputFormat);
+    if (inputSDFFormat != null) {
+      inputDateTimeFormatter = DateTimeFormat.forPattern(inputSDFFormat).withZoneUTC();
+    }
+    String outputSDFFormat = DateTimeFieldSpecUtils.getSDFPatternFromFormat(outputFormat);
+    if (outputSDFFormat != null) {
+      outputDateTimeFormatter = DateTimeFormat.forPattern(outputSDFFormat).withZoneUTC();
+    }
+  }
+
+  /**
+   * Transforms the given input value into the new format and bucket
+   * @param inputValue
+   * @return
+   */
+  public long transformDateTime(Object inputValue) {
+    Long dateTimeColumnValueMS = convertFromInputFormatToMillis(inputValue);
+    Long bucketedDateTimevalueMS = bucketDateTimeValueMS(dateTimeColumnValueMS);
+    Long dateTimeValueConverted = convertFromMillisToOutputFormat(bucketedDateTimevalueMS);
+    return dateTimeValueConverted;
+  }
+
+  /**
+   * Converts the input from inputFormat to millis
+   * @param dateTimeValue
+   * @return
+   */
+  private Long convertFromInputFormatToMillis(Object dateTimeValue) {
+
+    Long timeColumnValueMS = 0L;
+    if (dateTimeValue != null) {
```

Comment:
instead of checking this on a per record basis, setup epochToSDFConvertor and SDFToEpochConverter etc and pick the right converter in the beginning.